### PR TITLE
Combat analysis

### PR DESCRIPTION
**Pull request by ajlee21**
This PR is extending the current scripts to allow for multiple noise correction methods: limma, combat.  There are many files that changed but there are many duplicated changes.  The files you need to focus on are:

functions/generate_data_parallel.py
functions/pipeline.py

If you look at NB in Pseudomonas_analysis/Pa_sample_lvl_sim/
 "_uncorrected.ipynb", "_corrected.ipynb", "_create_figs_manuscript.ipynb" these changes are repeated in all other NBs so you don't need to review them all.

### CODE REVIEW

File: functions/generate_data_parallel.py [hunk 1]
```diff
@@ -886,13 +886,19 @@ def apply_correction_io(local_dir,
                 corrected_experiment_data = limma.removeBatchEffect(
                     experiment_data, batch=experiment_map)
 
+                # Convert R object to pandas df
```
Comment (arielah) [inline functions/generate_data_parallel.py]: Do you need to convert from R object to pandas dataframe twice? Does ComBat take in a dataframe and output an object?

Comment (arielah) [reply functions/generate_data_parallel.py]: Nevermind, my bad. These are in two separate blocks. That being said, it's worth considering pulling the object conversion out of the if/else statement. It's helpful to have only the things that are directly different in if/else statements, to make it clear what actually changes in the process by choosing different parameters and what stays the same.

File: configs/config_Pa_sample_combat.tsv [hunk 2]
```diff
@@ -0,0 +1,11 @@
+local_dir	"/home/alexandra/Documents/Data/Batch_effects/"
+dataset_name	"Pseudomonas_analysis"
+analysis_name	"Pa_sample_lvl_sim"
+NN_architecture	"NN_2500_30"
+num_simulated_samples	6000
+lst_num_experiments	[1, 2, 5, 10, 20, 50, 100, 500, 1000, 2000, 3000, 5000]
```
Comment (arielah) [inline configs/config_Pa_sample_combat.tsv]: Does this and configs/config_Pa_experiment_combat.tsv need to be separate files? They're identical except for two lines. Would it be possible to combine them into one file, maybe using different names for the num_simulated_samples=600 and num_simulated_samples=6000 terms?

**Comment by cgreene** [comment]
Big fan of DRY here :)

We used to have a bunch of copies of The Pragmatic Programmer floating around the lab.

**Comment by ajlee21** [comment]
@arielah and @ben-heil I have refactored the repository to get rid of redundancy.

The main changes can be found in:
functions/pipeline.py
functions/simulations.py 
ReadMe

File: functions/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+import gc
+from sklearn import preprocessing
+
+from joblib import Parallel, delayed
+import multiprocessing
+
+import warnings
+warnings.filterwarnings(action='ignore')
+
+sys.path.append("../../")
```
Comment (ben-heil) [inline functions/pipeline.py]:
```suggestion
```

Comment (ben-heil) [reply functions/pipeline.py]: This line is unnecessary since `pipeline.py` is already in the functions directory

File: functions/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+        sep='\t',
+        index_col=0)
+
+    print(rpkm_data.shape)
+
+    # 0-1 normalize per gene
+    rnaseq_scaled_df = preprocessing.MinMaxScaler().fit_transform(rpkm_data)
+    rnaseq_scaled_df = pd.DataFrame(rnaseq_scaled_df,
+                                    columns=rpkm_data.columns,
+                                    index=rpkm_data.index).T
```
Comment (ben-heil) [inline functions/pipeline.py]: Just making sure, you have genes as column names and samples as rows, right? Otherwise the normalization will be per-sample instead of per-gene

File: functions/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+
+    # Get associated experiment ids with expression data
+    experiment_ids_with_gene_expression = []
+
+    for experiment_id in experiment_ids:
+
+        if "Human" in dataset_name:
+            # Some project id values are descriptions
+            # We will skip these
+            if len(experiment_id) == 9:
```
Comment (ben-heil) [inline functions/pipeline.py]: Where does the number 9 come from here? Are you sure that it will be true for all experiment_ids?

File: functions/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+    train_architecture = params['NN_architecture']
+
+    # Read data
+    normalized_data = pd.read_table(
+        input_data_file,
+        header=0,
+        sep='\t',
+        index_col=0).T
+
+    print(normalized_data.shape)
```
Comment (ben-heil) [inline functions/pipeline.py]:
```suggestion
```

File: functions/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+    dataset_name = params['dataset_name']
+
+    # Read data
+    rpkm_data = pd.read_table(
+        raw_input_data_file,
+        header=0,
+        sep='\t',
+        index_col=0)
+
+    print(rpkm_data.shape)
```
Comment (ben-heil) [inline functions/pipeline.py]:
```suggestion
```

File: functions/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+
+    if "sample" in simulation_type:
+        num_simulated_samples = params["num_simulated_samples"]
+        lst_num_experiments = params["lst_num_experiments"]
+    else:
+        num_simulated_experiments = params["num_simulated_experiments"]
+        lst_num_partitions = params["lst_num_partitions"]
+
+    # Output files
+    base_dir = os.path.abspath(os.path.join(os.getcwd(), "../"))
```
Comment (ben-heil) [inline functions/pipeline.py]:
```suggestion
    base_dir = os.path.abspath(os.pardir)
```

Comment (ben-heil) [reply functions/pipeline.py]: I'm pretty sure this does what you want and also works across OSes

File: Human/README.md [hunk 9]
```diff
@@ -24,11 +24,11 @@ Given a compendium of gene expression partitions, how will increasing the number
...
 
-![Similarity_Human_sample](https://github.com/greenelab/simulate-expression-compendia/blob/master/results/Human_sample_lvl_sim_svcca.png)
+![Similarity_Human_sample](https://github.com/greenelab/simulate-expression-compendia/blob/master/Human/results/Human_sample_lvl_sim_svcca_limma.png)
 
 
 ## Experiment-level simulation experiment 
-Scripts can be found in [scripts/Human_experiment_lvl_sim](https://github.com/greenelab/simulate-expression-compendia/tree/master/Human_analysis/Human_experiment_lvl_sim)  
+Scripts to run this analysis can be found in [Human_experiment_lvl_sim.ipynb](https://github.com/greenelab/simulate-expression-compendia/blob/master/Human/Human_experiment_lvl_sim.ipynb)
```
Comment (arielah) [inline Human/README.md]: This is great! Having everything live in a single script is better than pointing to an entire directory, because then the order of scripts is abundantly clear.

File: functions/simulations.py [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    Returns
+    --------
+    similarity_score_df: df
+        Similarity scores for each number of experiment/partition added per run
+
+    permuted_scre: df
+        Similarity score comparing the permuted data to the simulated data per run
+    '''
+
+  # Main
```
Comment (ben-heil) [inline functions/simulations.py]:
```suggestion
```

File: functions/simulations.py [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+    --------
+    similarity_score_df: df
+        Similarity scores for each number of experiment/partition added per run
+
+    permuted_scre: df
+        Similarity score comparing the permuted data to the simulated data per run
+
+    '''
+
+  # Main
```
Comment (ben-heil) [inline functions/simulations.py]:
```suggestion
```

File: README.md [hunk 12]
```diff
@@ -49,38 +49,49 @@ In order to run this simulation on your own gene expression data the following s
 First we need to set up your local repository: 
 1. Clone the ```simulate-expression-compendia``` repository
 2. Set up conda environment using the command above
-3. Create a new analysis folder in the main directory (i.e. "NAME_analysis")
-4. Within your analysis folder add the same folders as found in ```Pseudomonas_analysis``` (i.e. experiment_lvl_sim/, sample_lvl_sim/, data/, logs/, models/)
-5. Copy contents of ```Pseudomonas_analysis/Pa_sample_lvl_sim/``` into your respective folder.  
-6. Copy contents of ```Pseudomonas_analysis/Pa_experiment_lvl_sim/``` into your respective folder.  
+3. Create a new analysis folder in the main directory. This is equivalent to the ```Pseudomonas``` directory
+4. Within your analysis folder create ```data/``` directory and ```input/```, ```metadata/``` subdirectories
```
[the remark this answers is not in the document]
Comment (ben-heil) [reply README.md]: Specifically the ones in step 4, to clarify

[the remark this answers is not in the document]
Comment (ajlee21) [reply README.md]: Technically the script can create those directories but then the script will halt because the data isn't found.  I think these are the minimum directories that the user will need to add to run the scripts

File: functions/pipeline.py (hunk 3 above)
Comment (ajlee21) [reply functions/pipeline.py]: Good catch

File: functions/pipeline.py (hunk 4 above)
Comment (ajlee21) [reply functions/pipeline.py]: Thank you for checking this.  That is correct, the normalization is performed per gene.  The reason for the transpose is because the downstream functions are expected to get a dataset that is gene x sample.  I'm not sure of a good way to check the dimension without consistent sample and gene naming.  There were earlier bugs when the dimensions were switched accidentally

File: functions/pipeline.py (hunk 5 above)
Comment (ajlee21) [reply functions/pipeline.py]: This was manually coded by me based looking at the metadata sheet.  This will not necessarily be true for all incoming data.  I should add a comment about this in the readme as something else for the user to change.

File: functions/pipeline.py (hunk 6 above)
Comment (ajlee21) [reply functions/pipeline.py]: Updated the print statement to act as a check for the user

**Comment by ajlee21** [comment]
Thank you both for making this repository a lot more user friendly and concise!

I realized that this PR was originally about this combat analysis, but one of the comments about this original PR was about the repo organization.  I didn't merge the original PR and instead just added the refactor changes ontop of this one which is why this PR is so large.  Next time I will definitely split accordingly.
